src/stream_base.h

```diff
@@ -146,10 +146,14 @@ class StreamResource {
                          const uv_buf_t* buf,
                          uv_handle_type pending,
                          void* ctx);
+  typedef void (*DestructCb)(void* ctx);
 
   StreamResource() : bytes_read_(0) {
   }
-  virtual ~StreamResource() = default;
+  virtual ~StreamResource() {
+    if (!destruct_cb_.is_empty())
+      destruct_cb_.fn(destruct_cb_.ctx);
```

Comment:
@bnoordhuis okay. so this took me far longer than I'd have liked but I now remember why I made the change this way to begin with.

`TLSWrap::stream_` points to the `StreamBase*` instance that's passed to `TLSWrap::TLSWrap()`. Each of these pointers are assigned to their own `FunctionTemplate` instance as an internal aligned pointer,  and the lifetime of each is independent from the other. Meaning `stream_` may be deleted even though the `TLSWrap` instance remains alive. In JS they're just attached as properties to each other.

So if the `StreamBase` instance is closed and released then `stream_` will point to invalid memory, even though the `TLSWrap` instance is still alive. Which means the necessary step is to zero out the `stream_` field in the `TLSWrap` instance, and also add `stream != nullptr` checks to the native methods like `TLSWrap::IsAlive()`, so it is no longer pointing to an invalid memory address.

To clarify (if that's possible) `TLSWrap` inherits from `StreamBase` and it holds a pointer to another `StreamBase` instance.

 I'll update the commit message with all this information.

FYI all that class template multiple inheritance stuff going on is what lead me to this solution in c0e6c668:
```cpp
#define ASSIGN_OR_RETURN_UNWRAP(ptr, obj, ...)                                \
  do {                                                                        \
    *ptr =                                                                    \
        Unwrap<typename node::remove_reference<decltype(**ptr)>::type>(obj);  \
    if (*ptr == nullptr)                                                      \
      return __VA_ARGS__;                                                     \
  } while (0)
```
In order to handle unwrapping in methods like `template<class Base, ...> StreamBase::JSMethod()` that need to convert a `void*` of a class instance in a template method to its parent class, that also has multiple inheritance.